pkg/code-generator/codegen/templates/resource_template.go

```diff
@@ -39,6 +40,22 @@ func (r *{{ .Name }}) SetMetadata(meta core.Metadata) {
 	r.Metadata = meta
 }
 
+func (r *{{ .Name }}) Hash() uint64 {
+	metaCopy := r.GetMetadata()
+	metaCopy.ResourceVersion = ""
+	return hashutils.HashAll(
+		metaCopy,
+{{- range .Fields }}
+	{{- if not ( or (eq .Name "metadata") (eq .Name "status") .IsOneof .SkipHashing ) }}
+		r.{{ upper_camel .Name }},
+	{{- end }}
+{{- end}}
+{{- range .Oneofs }}
+		r.{{ upper_camel .Name }},
+{{- end}}
+	)
+}
```

Comment:
here we are now generating a hashing function for every resource
this allows us to simplify the code in `Snapshot` as well as omit fields from hashing specified by the user (as a custom proto field option)